cmd/nodeJsBuild.go

```diff
@@ -0,0 +1,157 @@
+	return doublestar.Glob(pattern)
+}
+
+func (u *nodeJsBuildUtilsBundle) getwd() (dir string, err error) {
+	return os.Getwd()
+}
+
+func (u *nodeJsBuildUtilsBundle) dir(fileName string) string {
+	return path.Dir(fileName)
+}
+
+func (u *nodeJsBuildUtilsBundle) chdir(dir string) error {
+	return os.Chdir(dir)
+}
+
+func (u *nodeJsBuildUtilsBundle) getExecRunner() execRunner {
+	if u.execRunner == nil {
+		u.execRunner = &command.Command{}
+		u.execRunner.Stdout(log.Entry().Writer())
+		u.execRunner.Stderr(log.Entry().Writer())
+	}
+	return u.execRunner
+}
+
+func nodeJsBuild(config nodeJsBuildOptions, telemetryData *telemetry.CustomData) {
+	utils := nodeJsBuildUtilsBundle{}
+
+	err := runNodeJsBuild(&utils, &config)
+	if err != nil {
+		log.Entry().WithError(err).Fatal("step execution failed")
+	}
+}
+func runNodeJsBuild(utils nodeJsBuildUtilsInterface, options *nodeJsBuildOptions) error {
+	execRunner := utils.getExecRunner()
+	log.Entry().Infof("NPM Registry configuration: defaultNpmRegistry %s, sapNpmRegistry %s",
+		options.DefaultNpmRegistry, options.SapNpmRegistry)
+	environment := []string{"npm_config_@sap:registry=" + options.SapNpmRegistry}
+	if options.DefaultNpmRegistry != "" {
+		environment = append(environment, "npm_config_registry="+options.DefaultNpmRegistry)
+	}
+	execRunner.SetEnv(environment)
+
+	packageJSONFiles, err := findPackageJSONFiles(utils)
+	if err != nil {
+		return err
+	}
+
+	oldWorkingDirectory, err := utils.getwd()
+
+	for _, file := range packageJSONFiles {
+		dir := utils.dir(file)
+		_ = utils.chdir(dir)
+		packageLockExists, err := utils.fileExists("package-lock.json")
+
+		if err != nil {
+			return err
+		}
+		yarnLockExists, err := utils.fileExists("yarn.lock")
+		if err != nil {
+			return err
+		}
+		if options.Install {
+			err = installDependencies(dir, packageLockExists, yarnLockExists, execRunner)
+			if err != nil {
+				return err
+			}
+		}
+
+		for _, v := range options.RunScripts {
+			log.Entry().WithField("WorkingDirectory", dir).Info("run-script " + v)
+			err = execRunner.RunExecutable("npm", "run-script", v, "--if-present")
+			if err != nil {
+				return err
+			}
+		}
+		_ = utils.chdir(oldWorkingDirectory)
+	}
+
+	return err
+}
+
+func findPackageJSONFiles(utils nodeJsBuildUtilsInterface) ([]string, error) {
+	unfilteredListOfPackageJSONFiles, err := utils.glob("**/package.json")
+	if err != nil {
+		return nil, err
+	}
+
+	var packageJSONFiles []string
+
+	for _, file := range unfilteredListOfPackageJSONFiles {
+		if strings.Contains(file, "node_modules") {
+			continue
+		}
+		packageJSONFiles = append(packageJSONFiles, file)
+		log.Entry().Info("Discovered package.json file " + file)
+	}
+	return packageJSONFiles, nil
+}
+
+func installDependencies(dir string, packageLockExists bool, yarnLockExists bool, execRunner execRunner) (err error) {
+	log.Entry().WithField("WorkingDirectory", dir).Info("Running install")
+	if packageLockExists {
+		log.Entry().Info("npm ci")
```

Comment:
Log output should be removed in anticipation of https://github.com/SAP/jenkins-library/pull/1300.